src/adm/filter.rs

```diff
@@ -107,6 +119,90 @@ impl AdmFilter {
         l_sentry::report(&merged_tags, sentry::event_from_error(error));
     }
 
+    /// check to see if the bucket has been modified since the last time we updated.
+    pub async fn requires_update(&self) -> HandlerResult<bool> {
+        // don't update non-bucket versions (for now)
+        if !self.is_cloud() {
+            return Ok(false);
+        }
+        if let Some(bucket) = &self.source_url {
+            let host = bucket
+                .host()
+                .ok_or_else(|| {
+                    HandlerError::internal(&format!("Missing bucket Host {:?}", self.source))
+                })?
+                .to_string();
+
+            let obj = cloud_storage::Object::read(&host, bucket.path())
+                .await
+                .map_err(|e| {
+                    HandlerError::internal(&format!(
+                        "Could not read bucket {:?}, {:?}",
+                        self.source, e
+                    ))
+                })?;
+            if let Some(updated) = self.last_updated {
+                // if the bucket is older than when we last checked, do nothing.
+                return Ok(updated <= obj.updated);
+            };
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    /// Try to update the ADM filter data from the remote bucket.
+    pub async fn update(&mut self) -> HandlerResult<()> {
+        if let Some(bucket) = &self.source_url {
+            let adm_settings = AdmSettings::from_settings_bucket(bucket)
+                .await
+                .map_err(|e| {
+                    HandlerError::internal(&format!(
+                        "Invalid bucket data in {:?}: {:?}",
+                        self.source, e
+                    ))
+                })?;
+            for (adv, setting) in adm_settings.advertisers {
+                if setting.delete {
+                    trace!("Removing advertiser {:?}", &adv);
+                    self.filter_set.remove(&adv.to_lowercase());
+                };
+                trace!("Processing records for {:?}", &adv);
+                // DEFAULT included but sans special processing -- close enough
+                for country in &setting.include_regions {
+                    if !self.all_include_regions.contains(country) {
+                        self.all_include_regions.insert(country.clone());
+                    }
+                }
+                // map the settings to the URL we're going to be checking
+                self.filter_set.insert(adv.to_lowercase(), setting);
+            }
+            self.last_updated = Some(chrono::Utc::now());
+        }
+        Ok(())
+    }
+
+    // background settings updater.
+    pub async fn spawn_updater(&self) {
+        if !self.is_cloud() {
+            return;
+        }
+        let mut s = self.clone();
```

Comment:
This `clone`'s similar to the other one we removed in `server/mod.rs` -- we don't want it as it means the background updater is updating a separate copy of the `AdmFilter` that isn't otherwise used.

The background updater should switch to operating on I think that `Arc<RwLock<AdmFilter>>` as well.